specifyweb/frontend/js_src/lib/components/Interactions/LoanReturn.tsx

```diff
@@ -104,6 +104,36 @@ function PreparationReturn({
     ({ resolve, unresolved }) => resolve < unresolved
   );
 
+  const [bulkReturn, setBulkReturn] = React.useState(0);
+  const [bulkResolve, setBulkResolve] = React.useState(0);
+  const quantities = preparations.map(
+    (preparation) => preparation.get('quantity') ?? 0
+  );
+  const maxPrep = Math.max(...quantities);
+
+  const setQuantityToBulk = (
+    newCount: number,
+    type: 'resolve' | 'returns'
+  ): void => {
+    if (type === 'resolve') setBulkResolve(newCount);
+    else setBulkReturn(newCount);
+
+    setState((previousState) =>
+      previousState.map(({ resolve, returns, ...returnItem }, index) => {
+        const unresolved =
+          quantities[index] -
+          (preparations[index].get('quantityResolved') ?? 0);
+        const resolvedCount = Math.min(newCount, unresolved);
+
+        return {
+          ...returnItem,
+          resolve: type === 'resolve' ? resolvedCount : resolve,
+          returns: type === 'returns' ? resolvedCount : returns,
+        };
```

Comment:
this logic is almost correct, but not fully there:

see how if user modified returned amount, the resolved amount can get changed too:
https://github.com/specify/specify7/blob/xml-editor/specifyweb/frontend/js_src/lib/components/Interactions/PrepReturnRow.tsx#L109-L112

similarly, if user modified the resolved amount, the returned amount may change too - you need to add the same logic here (ideally you won't copy paste this logic between PrepReturnRow and here but would define a common small utility and reuse - even though this logic is quite concise, it's still a bit tricky, so better not have it defined in two places as that could lead to bugs like this)

Reply:
sure.

first, definitions:
- unresolved - total number of items that aren't accounted for yet - thus the maximum number of items we could account for in this transaction
- resolved - the number of items we want to account for. this of course has to be lower than unresolved count
- returned - the number of items we physically received back from the loan. this is not always the same as resolved count. i.e if a specimen was destoyed/lost/consumed, then it is resolved, but not returned. return count has to always be less or equal to the resolved count.

in your function at the moment, if I try to increase the return count, it may create a situation where return count is larger than resolved, which should not be allowed - in such cases, the resolved count should be increased too, to match the return count.
similarly, if your function is called to lower the resolved amount, it needs to make sure that returned amount is lowered to match resolved amount.


this is what the max and min in here is for:
https://github.com/specify/specify7/blob/xml-editor/specifyweb/frontend/js_src/lib/components/Interactions/PrepReturnRow.tsx#L109-L112
https://github.com/specify/specify7/blob/xml-editor/specifyweb/frontend/js_src/lib/components/Interactions/PrepReturnRow.tsx#L129-L132
those are making sure to enforce this constraint when user is modifying the returned/resolved count

now with two new fields you added in this PR, you do not have similar constraint enforcement yet. so my comment is that you need to handle these cases.
and since this logic is somewhat tricky and confusing, it would be best if you could define this logic in one place (some sort of function - or two functions, one for when resolved amount is changed, and another for when returned amount changed), and then use that function both when user modified a single row directly, and when user modified the field responsible for all rows.

Reply:
ok thank you very much. Will try that :+1: